File: gpio4j/src/main/java/org/overworld/gpio4j/serial/wire/Analog.java
package org.overworld.gpio4j.serial.wire;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

import org.overworld.gpio4j.serial.LineByte;
import org.overworld.gpio4j.serial.MessageType;
import org.overworld.gpio4j.serial.PinType;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.With;

@NoArgsConstructor
@AllArgsConstructor
@With
@Data
public class Analog implements Message {
	
	private Integer value;
	
	@Override
	public Analog parse(ByteArrayInputStream wireData) throws IOException {
		
		if (wireData.available() < 2) throw new IllegalStateException("Incorrect number of bytes");
		
		value = new LineByte(wireData.readNBytes(2)).asInt();
		
		return this;
	}

	@Override
	public ByteArrayOutputStream wireData() {

		if (value == null) throw new IllegalStateException("Value is not set when generating wire data");
		
		ByteArrayOutputStream result = new ByteArrayOutputStream();
		result.writeBytes(new LineByte(value).getHiLo());
		return result;
	}

	@Override
	public MessageType getType() {
		return MessageType.ANALOG;
	}
}

File: gpioBoardDriver/serial/Analog.cpp
#include "Analog.h"

class Analog : public Message {
public:
    Analog() : value(0) {}
    Analog(int value) : value(value) {}

    Analog* parse(std::istream& wireData) override {
        if (wireData.rdbuf()->in_avail() < 2) throw std::runtime_error("Incorrect number of bytes");

        uint8_t bytes[2];
        wireData.read(reinterpret_cast<char*>(bytes), 2);
        value = LineByte((bytes[0] << 8) | bytes[1]).asInt();

        return this;
    }

    std::vector<uint8_t> wireData() override {
        if (!value) throw std::runtime_error("Value is not set when generating wire data");

        LineByte lineByte(value);
        return lineByte.getHiLo();
    }

    MessageType getType() override {
        return MessageType::ANALOG;
    }

private:
    int value;
};
